### PR TITLE
Fix #3540: Implicit class with two arguments does not fail compilation

### LINKED ISSUE #3540: Implicit class with two arguments does not fail compilation
```scala
object Test {
  implicit class Foo(i: Int, s: String)
}
```
Fails in `scalac` with
```
Foo.scala:2: error: implicit classes must accept exactly one primary constructor parameter
  implicit class Foo(i: Int, s: String)
                 ^
one error found
```

But not on `dotc`. This error should probably be emitted when desugaring the `implicit class`.

### PR DESCRIPTION
**Pull request by vitorsvieira**
This PR aims #3540 and #2464.

Where:

```
object Test {
  implicit class Foo(i: Int, s: String)
}
```

And

```
object Foo {
  object Bar
  implicit class Bar
}
```

Compilation must fail providing `Implicit classes must accept exactly one primary constructor parameter` error message and further explanation.

### CODE REVIEW

File: tests/neg/i3540.scala [hunk 1]
```diff
@@ -0,0 +1,3 @@
+object Test {
+  implicit class Foo(i: Int, s: String) // error: Implicit classes must accept exactly one primary constructor parameter
+}
```
Comment (allanrenucci) [inline tests/neg/i3540.scala]: Can you add more tests here:
```scala
implicit class Foo0
implicit class Foo1()
implicit class Foo2()(x: Int)
```

Comment (nicolasstucki) [reply tests/neg/i3540.scala]: You should also add 
```scala
implicit case class Bar0
implicit case class Bar1()
implicit case class Bar2()(x: Int)
```

Comment (vitorsvieira) [reply tests/neg/i3540.scala]: OK. Adding these tests,

Comment (nicolasstucki) [reply tests/neg/i3540.scala]: Probably

Comment (allanrenucci) [reply tests/neg/i3540.scala]: You simply need to add `// error` next to the lines that do not compile

Comment (vitorsvieira) [reply tests/neg/i3540.scala]: 👍 Added.

Comment (allanrenucci) [reply tests/neg/i3540.scala]: Can you also add the error messages emitted by the compiler:
```scala
implicit class Foo0 // error: <the error message here>
```
You can get the error messages by running in sbt:
```
dotc tests/neg/i3540.scala
```

Comment (vitorsvieira) [reply tests/neg/i3540.scala]:
```

Comment (vitorsvieira) [reply tests/neg/i3540.scala]: Added error messages to the comments.

File: compiler/src/dotty/tools/dotc/reporting/diagnostic/messages.scala [hunk 2]
```diff
@@ -453,6 +452,33 @@ object messages {
...
+          |
+          |$allowedImplicit1
+          |
+          |Or as:
+          |
+          |$allowedImplicit2
+          |
+          |It’s possible to create an implicit class with more than one
+          |non-implicit argument, but such classes are not used during implicit lookup.
+          |""" + implicitClassRestrictionsText
```
Comment (allanrenucci) [inline compiler/src/dotty/tools/dotc/reporting/diagnostic/messages.scala]: I would say instead:
```scala
val explanation = {
  val example = "implicit class RichDate(date: java.util.Date)"
  hl"""Implicit classes may only take one non-implicit argument in their constructor. For example:
      |
      | $example
      |
      |While it’s possible to create an implicit class with more than one non-implicit argument,
      |such classes aren’t used during implicit lookup.
      |""" + implicitClassRestrictionsText
}
```

File: compiler/test/dotty/tools/dotc/reporting/ErrorMessagesTests.scala [hunk 3]
```diff
@@ -801,6 +801,23 @@ class ErrorMessagesTests extends ErrorMessagesTest {
...
+        |  implicit class Foo(i: Int, s: String)
+        |}
+      """.stripMargin
+    }
+    .expect { (itcx, messages) =>
+      implicit val ctx: Context = itcx
+
+      assertMessageCount(1, messages)
+      val err :: Nil = messages
+
```
Comment (allanrenucci) [inline compiler/test/dotty/tools/dotc/reporting/ErrorMessagesTests.scala]: Remove this white line

File: tests/neg/i3540.scala [hunk 4]
```diff
@@ -0,0 +1,9 @@
+object Test {
+  implicit class Foo(i: Int, s: String) // error: Implicit classes must accept exactly one primary constructor parameter
+  implicit class Foo0                   // error: Implicit classes must accept exactly one primary constructor parameter
+  implicit class Foo1()                 // error: Implicit classes must accept exactly one primary constructor parameter
+  implicit class Foo2()(x: Int)         // error: Implicit classes must accept exactly one primary constructor parameter
+  implicit case class Bar0              // error: A case class must have at least one parameter list
+  implicit case class Bar1()            // error: A case class may not be defined as implicit
+  implicit case class Bar2()(x: Int)    // error: A case class may not be defined as implicit
+}
```
Comment (allanrenucci) [inline tests/neg/i3540.scala]: Add new line at the end of a file

File: compiler/src/dotty/tools/dotc/reporting/diagnostic/messages.scala (hunk 2 above)
Comment (vitorsvieira) [reply compiler/src/dotty/tools/dotc/reporting/diagnostic/messages.scala]: Perfect! Updated.

File: tests/neg/i3540.scala (hunk 4 above)
Comment (vitorsvieira) [reply tests/neg/i3540.scala]: Added.

**Pull request merged**